### PR TITLE
feat: add option to sigma-clip sub-bands homogeneously

### PR DESCRIPTION
**Pull request by steven-murray**
This adds several options to a new `sigma_clip()` function, including `clip_type` which specifies whether data should be directly clipped or accumulated over some axis before assessing the threshold. A placeholder for allowing the expected variance to be passed has also been added, but not piped through (i.e. you can't specify this on the CLI yet).

The reason that the expected variance can't be piped through easily is that there's no obvious way of getting the auto spectra that works for both redundantly averaged and non-averaged data input.

### CODE REVIEW

**Review by jsdillon — CHANGES_REQUESTED** [review]
Mostly minor comments, though I think think it'd be nice to add a unit test for the new modes.

Also, a bunch of unit tests are failing... so that's fun.

File: hera_cal/lstbin_simple.py [hunk 1]
```diff
@@ -501,6 +502,114 @@ def threshold_flags(
...
+        ``mean`` or ``median``. In this case, if for example a 2D array is passed in
+        and ``threshold_axis=1`` (but no ``flag_bands`` is passed), then the mean of
+        the absolute zscores is take along the final axis, and the output flags are
+        applied homogeneously across this axis based on this mean.
+    clip_type
+        The type of sigma clipping to perform. If ``direct``, each datum is flagged
+        individually. If ``mean`` or ``median``, an entire sub-band of the data is
+        flagged if its mean (absolute) zscore is beyond the threshold.
+    flag_bands
+        A list of tuples specifying the start and end indices of the threshold axis
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_cal/lstbin_simple.py:545]: Good idea. Done.

File: hera_cal/lstbin_simple.py [hunk 2]
```diff
@@ -501,6 +502,114 @@ def threshold_flags(
...
+        array = np.array(array)
+
+    # ensure array passes min_N criterion:
+    if array.shape[median_axis] < min_N:
+        return np.zeros_like(array, dtype=bool)
+
+    if not isinstance(array, np.ma.MaskedArray):
+        array = np.ma.MaskedArray(array, mask=np.nonzero(np.isnan(array)))
+
+    location = np.expand_dims(np.ma.median(array, axis=median_axis), axis=median_axis)
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_cal/lstbin_simple.py:566]: I think you need to use `np.ma.median` on a `MaskedArray`, which is just what we've ensured `array` to be.

File: hera_cal/lstbin_simple.py [hunk 3]
```diff
@@ -501,6 +502,114 @@ def threshold_flags(
...
+        elif clip_type == 'mean':
+            # In this mode, an entire sub-band of the data is flagged if its mean
+            # (absolute) zscore is beyond the threshold.
+            mean_abs_dev = np.mean(subz, axis=threshold_axis)
+            subflags[:] = np.expand_dims(mean_abs_dev > threshold, axis=threshold_axis)
+        elif clip_type == 'median':
+            # In this mode, an entire sub-band of the data is flagged if its median
+            # (absolute) zscore is beyond the threshold.
+            mean_abs_dev = np.median(subz, axis=threshold_axis)
+            subflags[:] = np.expand_dims(mean_abs_dev > threshold, axis=threshold_axis)
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_cal/lstbin_simple.py:603]: Good point. Done.

File: hera_cal/lstbin_simple.py [hunk 4]
```diff
@@ -541,6 +653,17 @@ def lst_average(
         The minimum number of unflagged samples required to perform sigma clipping.
     flag_below_min_N
         Whether to flag data that has fewer than ``sigma_clip_min_N`` unflagged samples.
+    sigma_clip_subbands
+        A list of integers specifying the start and end indices of the frequency axis
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_cal/lstbin_simple.py:657]: Good point, done.

File: hera_cal/lstbin_simple.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsdillon) [reply hera_cal/lstbin_simple.py:574]: Oh, I missed that. Sounds good.

File: hera_cal/lstbin_simple.py [hunk 5]
```diff
@@ -268,6 +269,9 @@ def reduce_lst_bins(
     mutable: bool = False,
     sigma_clip_thresh: float | None = None,
     sigma_clip_min_N: int = 4,
+    sigma_clip_type: str = 'direct',
+    sigma_clip_subbands: list[int] | None = None,
+    sigma_clip_scale: list[np.ndarray] | None = None,
```
Comment (jsdillon) [inline hera_cal/lstbin_simple.py:274]: These new features need documenting

File: scripts/lstbin_simple.py [hunk 6]
```diff
@@ -60,6 +60,8 @@
 
 kwargs['output_flagged'] = not kwargs.pop('no_flagged_mode')
 kwargs['output_inpainted'] = kwargs.pop("do_inpaint_mode")
+kwargs['sigma_clip_subbands'] = [int(b) for b in kwargs["sigma_clip_subbands"].split(",")]
```
Comment (jsdillon) [inline scripts/lstbin_simple.py:63]: Has this been added to the argparser?

Also, this raises an interesting question about the best way to specify subbands from the commandline and/or from a TOML. 

I'm using this format, which is in MHz. It's similar to something Aaron EW did for H4C, though I think he used channels. It's a lot more human readable, but it's a bit clunky:

https://github.com/HERA-Team/hera_pipelines/blob/46a6018bf497d1066fd242aaf01fa6e0a2aaa9fd/pipelines/h6c/idr2/v2/post-lstbin/h6c_pspec.toml#L15

**Review by jsdillon — CHANGES_REQUESTED** [review]
Just a couple of questions/comments, but nearly good to go!

File: scripts/lstbin_simple.py (hunk 6 above)
Comment (steven-murray) [reply scripts/lstbin_simple.py:63]: Yes, it's been added to the argparser. The format I'm using is like:

```
--sigma-clip-subbands 0,150,1000
```

This assumes that *every* channel belongs to a subband, so you just specify the edges. I could make it more general, using the tilde notation, but I'm not perfectly sure it's necessary in this case?

Comment (jsdillon) [reply scripts/lstbin_simple.py:63]: I'd like to use the tilde notation. It's more standard and it's not necessarily true that every channel belongs to a subband (e.g. FM).

Comment (steven-murray) [reply scripts/lstbin_simple.py]: OK, I've done this now.

**Review by jsdillon — APPROVED** [review]
Looks good. Thanks @steven-murray !

**Pull request merged**